packages/components/src/popover/index.js

```diff
@@ -485,97 +480,38 @@ const Popover = (
-				// Close on escape.
-				if ( event.keyCode === ESCAPE && onClose ) {
-					event.stopPropagation();
-					onClose();
-				}
-			}
-
-			node.addEventListener( 'keydown', maybeClose );
-
-			return () => {
-				node.removeEventListener( 'keydown', maybeClose );
-			};
-		},
-		[ onClose ]
-	);
-
-	const constrainedTabbingRef = useConstrainedTabbing();
-	const focusReturnRef = useFocusReturn();
-	const focusOnMountRef = useFocusOnMount( focusOnMount );
-	const focusOutsideProps = useFocusOutside( handleOnFocusOutside );
-	const mergedRefs = useMergeRefs( [
-		ref,
-		containerRef,
-		// Don't register the event at all if there's no onClose callback.
-		onClose ? closeEventRef : null,
-		focusOnMount ? constrainedTabbingRef : null,
-		focusOnMount ? focusReturnRef : null,
-		focusOnMount ? focusOnMountRef : null,
-	] );
-
-	/**
-	 * Shims an onFocusOutside callback to be compatible with a deprecated
-	 * onClickOutside prop function, if provided.
-	 *
-	 * @param {FocusEvent} event Focus event from onFocusOutside.
-	 */
-	function handleOnFocusOutside( event ) {
-		// Defer to given `onFocusOutside` if specified. Call `onClose` only if
-		// both `onFocusOutside` and `onClickOutside` are unspecified. Doing so
-		// assures backwards-compatibility for prior `onClickOutside` default.
-		if ( onFocusOutside ) {
+	const onDialogClose = ( type, event ) => {
+		// Ideally the popover should have just a single onClose prop and
+		// not three props that potentially do the same thing.
+		if ( type === 'focus-outside' && onFocusOutside ) {
 			onFocusOutside( event );
-			return;
-		} else if ( ! onClickOutside ) {
-			if ( onClose ) {
-				onClose();
-			}
-			return;
-		}
-
-		// Simulate MouseEvent using FocusEvent#relatedTarget as emulated click
-		// target. MouseEvent constructor is unsupported in Internet Explorer.
-		let clickEvent;
-		try {
-			clickEvent = new window.MouseEvent( 'click' );
-		} catch ( error ) {
-			clickEvent = document.createEvent( 'MouseEvent' );
-			clickEvent.initMouseEvent(
-				'click',
-				true,
-				true,
-				window,
-				0,
-				0,
-				0,
-				0,
-				0,
-				false,
-				false,
-				false,
-				false,
-				0,
-				null
-			);
+		} else if ( type === 'focus-outside' && onClickOutside ) {
+			// Simulate MouseEvent using FocusEvent#relatedTarget as emulated click
+			// target. MouseEvent constructor is unsupported in Internet Explorer.
+			const clickEvent = new window.MouseEvent( 'click' );
+
+			Object.defineProperty( clickEvent, 'target', {
+				get: () => event.relatedTarget,
+			} );
+
+			deprecated( 'Popover onClickOutside prop', {
+				since: '5.3',
+				alternative: 'onFocusOutside',
+			} );
+
+			onClickOutside( clickEvent );
+		} else if ( onClose ) {
+			onClose();
 		}
+	};
 
-		Object.defineProperty( clickEvent, 'target', {
-			get: () => event.relatedTarget,
-		} );
+	const [ dialogRef, dialogProps ] = useDialog( {
+		focusOnMount,
+		__unstableOnClose: onDialogClose,
+		onClose: onDialogClose,
```

Comment:
What's the difference between these two?

Reply:
I also mean: do we need to set both here?

Reply:
Ideally we should remove `__unstableOnClose` it's there because Popover need to differentiate between the closing when it happens on focus out and the closing when you use escape... Dialogs in general don't care and if we do your suggestion (merge onFocusOutside and onClose) we should be able to remove the unstable one